File: btc_predictor/backfill_gaps.js
/**
 * Backfill Data Gaps - TradingView
 *
 * Fetches historical data to fill gaps when the collector was not running.
 * TradingView provides ~300 historical candles when connecting to a chart.
 *
 * Usage:
 *   node backfill_gaps.js           - Backfill all assets
 *   node backfill_gaps.js Crypto    - Backfill specific category
 *   node backfill_gaps.js Crypto BTC - Backfill specific asset
 */

const TradingView = require('@mathieuc/tradingview');
const fs = require('fs');
const path = require('path');

const ASSETS = {
    Crypto: {
        BTC: 'COINBASE:BTCUSD',
        ETH: 'COINBASE:ETHUSD',
        SOL: 'COINBASE:SOLUSD',
        XRP: 'COINBASE:XRPUSD',
        DOGE: 'COINBASE:DOGEUSD',
        ADA: 'COINBASE:ADAUSD',
        AVAX: 'COINBASE:AVAXUSD',
        DOT: 'COINBASE:DOTUSD',
        POL: 'COINBASE:POLUSD',
        LINK: 'COINBASE:LINKUSD',
        LTC: 'COINBASE:LTCUSD',
        UNI: 'COINBASE:UNIUSD',
        ATOM: 'COINBASE:ATOMUSD',
        XLM: 'COINBASE:XLMUSD',
        ALGO: 'COINBASE:ALGOUSD',
    },
    "Stock Market": {
        SPY: 'AMEX:SPY',
        QQQ: 'NASDAQ:QQQ',
        DIA: 'AMEX:DIA',
        IWM: 'AMEX:IWM',
        AAPL: 'NASDAQ:AAPL',
        MSFT: 'NASDAQ:MSFT',
        GOOGL: 'NASDAQ:GOOGL',
        AMZN: 'NASDAQ:AMZN',
        NVDA: 'NASDAQ:NVDA',
        META: 'NASDAQ:META',
        TSLA: 'NASDAQ:TSLA',
        JPM: 'NYSE:JPM',
        V: 'NYSE:V',
        MA: 'NYSE:MA',
        BAC: 'NYSE:BAC',
        UNH: 'NYSE:UNH',
        JNJ: 'NYSE:JNJ',
        WMT: 'NYSE:WMT',
        PG: 'NYSE:PG',
        XOM: 'NYSE:XOM',
    },
    Commodities: {
        GOLD: 'TVC:GOLD',
        SILVER: 'TVC:SILVER',
        OIL: 'TVC:USOIL',
        NATGAS: 'NYMEX:NG1!',
        COPPER: 'COMEX:HG1!',
        PLATINUM: 'TVC:PLATINUM',
    },
    Currencies: {
        EURUSD: 'FX:EURUSD',
        GBPUSD: 'FX:GBPUSD',
        USDJPY: 'FX:USDJPY',
        AUDUSD: 'FX:AUDUSD',
        USDCAD: 'FX:USDCAD',
        USDCHF: 'FX:USDCHF',
        NZDUSD: 'FX:NZDUSD',
        EURGBP: 'FX:EURGBP',
    },
};

const DATA_DIR = path.join(__dirname, '..', 'Data');

function ensureDir(dirPath) {
    if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath, { recursive: true });
    }
}

function getWeekStart(date) {
    const d = new Date(date);
    const day = d.getUTCDay();
    const diff = d.getUTCDate() - day + (day === 0 ? -6 : 1);
    d.setUTCDate(diff);
    return d;
}

function formatDate(date) {
    return date.toISOString().split('T')[0];
}

// Load existing candles for a day
function loadDayCandles(category, asset, dateStr) {
    const assetDir = path.join(DATA_DIR, category, asset);
    const weekDirs = fs.readdirSync(assetDir).filter(d => d.startsWith('week_'));

    for (const weekDir of weekDirs) {
        const dayFile = path.join(assetDir, weekDir, `${dateStr}.json`);
        if (fs.existsSync(dayFile)) {
            try {
                return JSON.parse(fs.readFileSync(dayFile, 'utf8'));
            } catch {
                return [];
            }
        }
    }
    return [];
}

// Save candles to appropriate day file
function saveDayCandles(category, asset, dateStr, candles) {
    if (candles.length === 0) return;

    const date = new Date(dateStr);
    const weekStart = getWeekStart(date);
    const assetDir = path.join(DATA_DIR, category, asset);
    const weekFolder = path.join(assetDir, `week_${formatDate(weekStart)}`);
    ensureDir(weekFolder);

    const dayFile = path.join(weekFolder, `${dateStr}.json`);

    // Sort and deduplicate
    const unique = {};
    for (const c of candles) {
        unique[c.timestamp] = c;
    }
    const sorted = Object.values(unique).sort((a, b) => a.timestamp - b.timestamp);

    fs.writeFileSync(dayFile, JSON.stringify(sorted, null, 2));
    return sorted.length;
}

// Fetch historical data for an asset
function fetchHistoricalData(client, category, assetName, symbol) {
    return new Promise((resolve) => {
        const chart = new client.Session.Chart();

        // Request more historical data
        chart.setMarket(symbol, {
            timeframe: '1',
            range: 300, // Request 300 candles (~5 hours of 1-min data)
        });

        let resolved = false;
        const timeout = setTimeout(() => {
            if (!resolved) {
                resolved = true;
                chart.delete();
                resolve({ success: false, error: 'timeout', candles: [] });
            }
        }, 15000);

        chart.onUpdate(() => {
            if (resolved) return;

            const periods = chart.periods;
            if (!periods || periods.length === 0) return;

            const candles = periods.map(p => ({
                timestamp: p.time * 1000,
                open: p.open,
                high: p.max,
                low: p.min,
                close: p.close,
                volume: p.volume || 0
            }));

            clearTimeout(timeout);
            resolved = true;
            chart.delete();
            resolve({ success: true, candles });
        });

        chart.onError((err) => {
            if (!resolved) {
                clearTimeout(timeout);
                resolved = true;
                chart.delete();
                resolve({ success: false, error: err, candles: [] });
            }
        });
    });
}

// Main backfill function
async function backfillAsset(client, category, assetName, symbol) {
    console.log(`\nBackfilling ${category}/${assetName}...`);

    const result = await fetchHistoricalData(client, category, assetName, symbol);

    if (!result.success || result.candles.length === 0) {
        console.log(`  Failed: ${result.error || 'no data'}`);
        return { success: false, filled: 0 };
    }

    console.log(`  Received ${result.candles.length} historical candles`);

    // Group candles by day
    const byDay = {};
    for (const candle of result.candles) {
        const dayStr = formatDate(new Date(candle.timestamp));
        if (!byDay[dayStr]) byDay[dayStr] = [];
        byDay[dayStr].push(candle);
    }

    let totalFilled = 0;

    // Merge with existing data for each day
    for (const [dayStr, newCandles] of Object.entries(byDay)) {
        const existing = loadDayCandles(category, assetName, dayStr);
        const existingTimestamps = new Set(existing.map(c => c.timestamp));

        // Find candles we don't have
        const toAdd = newCandles.filter(c => !existingTimestamps.has(c.timestamp));

        if (toAdd.length > 0) {
            const merged = [...existing, ...toAdd];
            const saved = saveDayCandles(category, assetName, dayStr, merged);
            console.log(`  ${dayStr}: Added ${toAdd.length} candles (total: ${saved})`);
            totalFilled += toAdd.length;
        }
    }

    if (totalFilled === 0) {
        console.log(`  No gaps to fill`);
    }

    return { success: true, filled: totalFilled };
}

async function main() {
    const args = process.argv.slice(2);
    const filterCategory = args[0];
    const filterAsset = args[1];

    console.log('============================================================');
    console.log('TradingView Data Backfill');
    console.log('============================================================');
    console.log(`Data directory: ${DATA_DIR}`);
    if (filterCategory) console.log(`Category filter: ${filterCategory}`);
    if (filterAsset) console.log(`Asset filter: ${filterAsset}`);
    console.log('============================================================\n');

    const client = new TradingView.Client();
    let totalFilled = 0;
    let assetsProcessed = 0;

    for (const [category, assets] of Object.entries(ASSETS)) {
        if (filterCategory && category !== filterCategory) continue;

        for (const [assetName, symbol] of Object.entries(assets)) {
            if (filterAsset && assetName !== filterAsset) continue;

            try {
                const result = await backfillAsset(client, category, assetName, symbol);
                if (result.success) {
                    totalFilled += result.filled;
                }
                assetsProcessed++;
            } catch (err) {
                console.log(`  Error: ${err.message}`);
            }

            // Small delay between assets
            await new Promise(r => setTimeout(r, 500));
        }
    }

    client.end();

    console.log('\n============================================================');
    console.log('Backfill Complete');
    console.log(`Assets processed: ${assetsProcessed}`);
    console.log(`Total candles filled: ${totalFilled}`);
    console.log('============================================================');
}

main().catch(err => {
    console.error('Error:', err.message);
    process.exit(1);
});

File: btc_predictor/backfill_gaps.py
"""
Data Gap Checker and Backfill Script
Detects gaps in collected data and fills them automatically.
Run this on startup to ensure continuous data coverage.

Usage:
    python backfill_gaps.py [--check-only]
"""

import os
import sys
import json
import time
import requests
from pathlib import Path
from datetime import datetime, timedelta
from collections import defaultdict

DATA_DIR = Path(__file__).parent.parent / 'Data'

# Coinbase API for crypto backfill
COINBASE_API = 'https://api.exchange.coinbase.com'

# Assets that can be backfilled via Coinbase
COINBASE_ASSETS = {
    'BTC': 'BTC-USD',
    'ETH': 'ETH-USD',
    'SOL': 'SOL-USD',
    'XRP': 'XRP-USD',
    'DOGE': 'DOGE-USD',
    'ADA': 'ADA-USD',
    'AVAX': 'AVAX-USD',
    'DOT': 'DOT-USD',
    'MATIC': 'MATIC-USD',
    'LINK': 'LINK-USD',
    'UNI': 'UNI-USD',
    'ATOM': 'ATOM-USD',
    'LTC': 'LTC-USD',
    'XLM': 'XLM-USD',
}


def get_week_number(timestamp_ms):
    """Get week number from timestamp"""
    dt = datetime.fromtimestamp(timestamp_ms / 1000)
    return dt.isocalendar()[1]


def load_all_candles(asset_dir):
    """Load all candles for an asset and return sorted by timestamp"""
    all_candles = []

    if not asset_dir.exists():
        return []

    for week_dir in asset_dir.iterdir():
        if not week_dir.name.startswith('week_'):
            continue
        for json_file in week_dir.glob('*.json'):
            try:
                with open(json_file, 'r') as f:
                    data = json.load(f)
                    all_candles.extend(data)
            except (json.JSONDecodeError, IOError):
                continue

    # Dedupe and sort
    unique = {c['timestamp']: c for c in all_candles}
    return sorted(unique.values(), key=lambda x: x['timestamp'])


def find_gaps(candles, max_gap_minutes=5):
    """Find gaps in candle data larger than max_gap_minutes"""
    gaps = []

    if len(candles) < 2:
        return gaps

    for i in range(1, len(candles)):
        prev_ts = candles[i-1]['timestamp']
        curr_ts = candles[i]['timestamp']
        gap_minutes = (curr_ts - prev_ts) / 60000

        if gap_minutes > max_gap_minutes:
            gaps.append({
                'start': prev_ts,
                'end': curr_ts,
                'minutes': int(gap_minutes),
                'start_dt': datetime.fromtimestamp(prev_ts / 1000).isoformat(),
                'end_dt': datetime.fromtimestamp(curr_ts / 1000).isoformat(),
            })

    return gaps


def fetch_coinbase_candles(product_id, start_ts, end_ts):
    """Fetch candles from Coinbase to fill a gap"""
    candles = []

    # Coinbase returns max 300 candles per request
    # Work backwards from end to start
    current_end = end_ts

    while current_end > start_ts:
        # Calculate start for this batch (max 300 minutes back)
        batch_start = max(start_ts, current_end - (300 * 60 * 1000))

        url = f'{COINBASE_API}/products/{product_id}/candles'
        params = {
            'start': datetime.fromtimestamp(batch_start / 1000).isoformat(),
            'end': datetime.fromtimestamp(current_end / 1000).isoformat(),
            'granularity': 60,  # 1 minute
        }

        try:
            response = requests.get(url, params=params, timeout=10)
            if response.status_code == 200:
                data = response.json()
                for row in data:
                    # Coinbase format: [timestamp, low, high, open, close, volume]
                    candles.append({
                        'timestamp': row[0] * 1000,  # Convert to ms
                        'open': float(row[3]),
                        'high': float(row[2]),
                        'low': float(row[1]),
                        'close': float(row[4]),
                        'volume': float(row[5]),
                    })
            else:
                print(f'  Warning: Coinbase returned {response.status_code}')
                break
        except Exception as e:
            print(f'  Error fetching from Coinbase: {e}')
            break

        current_end = batch_start
        time.sleep(0.1)  # Rate limiting

    return candles


def save_candles(asset_dir, candles):
    """Save candles to appropriate week folders"""
    if not candles:
        return 0

    # Group by week
    by_week = defaultdict(list)
    for candle in candles:
        week = get_week_number(candle['timestamp'])
        by_week[week].append(candle)

    saved = 0
    for week, week_candles in by_week.items():
        week_dir = asset_dir / f'week_{week:02d}'
        week_dir.mkdir(parents=True, exist_ok=True)

        # Use timestamp range for filename
        min_ts = min(c['timestamp'] for c in week_candles)
        max_ts = max(c['timestamp'] for c in week_candles)
        filename = f'backfill_{min_ts}_{max_ts}.json'

        filepath = week_dir / filename
        with open(filepath, 'w') as f:
            json.dump(sorted(week_candles, key=lambda x: x['timestamp']), f)

        saved += len(week_candles)

    return saved


def check_and_backfill_asset(category, asset, check_only=False):
    """Check for gaps and optionally backfill them"""
    asset_dir = DATA_DIR / category / asset
    candles = load_all_candles(asset_dir)

    if not candles:
        print(f'  No data found')
        return {'gaps': 0, 'filled': 0}

    # Get date range
    first_dt = datetime.fromtimestamp(candles[0]['timestamp'] / 1000)
    last_dt = datetime.fromtimestamp(candles[-1]['timestamp'] / 1000)

    print(f'  Data range: {first_dt.strftime("%Y-%m-%d %H:%M")} to {last_dt.strftime("%Y-%m-%d %H:%M")}')
    print(f'  Total candles: {len(candles):,}')

    # Check for gap between last candle and now
    now_ts = int(datetime.now().timestamp() * 1000)
    gap_to_now = (now_ts - candles[-1]['timestamp']) / 60000

    if gap_to_now > 5:
        print(f'  Gap to current time: {int(gap_to_now)} minutes')

    # Find internal gaps
    gaps = find_gaps(candles)

    if gaps:
        print(f'  Found {len(gaps)} gap(s):')
        for gap in gaps[:5]:  # Show first 5
            print(f'    - {gap["minutes"]} min gap: {gap["start_dt"]} to {gap["end_dt"]}')
        if len(gaps) > 5:
            print(f'    ... and {len(gaps) - 5} more')
    else:
        print(f'  No gaps found (data is continuous)')

    # Backfill if requested and asset is supported
    filled = 0
    if not check_only and gaps:
        coinbase_symbol = COINBASE_ASSETS.get(asset)
        if coinbase_symbol:
            print(f'  Backfilling from Coinbase...')
            for gap in gaps:
                new_candles = fetch_coinbase_candles(
                    coinbase_symbol,
                    gap['start'],
                    gap['end']
                )
                if new_candles:
                    saved = save_candles(asset_dir, new_candles)
                    filled += saved
                    print(f'    Filled {saved} candles for gap at {gap["start_dt"]}')
        else:
            print(f'  Cannot backfill (no API source for {asset})')

    # Also try to fill gap to current time for crypto
    if not check_only and gap_to_now > 5:
        coinbase_symbol = COINBASE_ASSETS.get(asset)
        if coinbase_symbol:
            print(f'  Backfilling gap to current time...')
            new_candles = fetch_coinbase_candles(
                coinbase_symbol,
                candles[-1]['timestamp'],
                now_ts
            )
            if new_candles:
                saved = save_candles(asset_dir, new_candles)
                filled += saved
                print(f'    Filled {saved} candles up to now')

    return {'gaps': len(gaps), 'filled': filled}


def main():
    check_only = '--check-only' in sys.argv

    print('=' * 60)
    print('S.U.P.I.D. Data Gap Checker')
    print('=' * 60)
    print(f'Mode: {"Check Only" if check_only else "Check and Backfill"}')
    print(f'Data directory: {DATA_DIR}')
    print()

    total_gaps = 0
    total_filled = 0

    # Scan all categories
    for category_dir in DATA_DIR.iterdir():
        if not category_dir.is_dir():
            continue

        category = category_dir.name
        print(f'\n[{category}]')

        for asset_dir in category_dir.iterdir():
            if not asset_dir.is_dir():
                continue

            asset = asset_dir.name
            print(f'\n{asset}:')

            result = check_and_backfill_asset(category, asset, check_only)
            total_gaps += result['gaps']
            total_filled += result['filled']

    print()
    print('=' * 60)
    print(f'Summary: Found {total_gaps} gaps across all assets')
    if not check_only:
        print(f'Filled: {total_filled} candles')
    print('=' * 60)


if __name__ == '__main__':
    main()
